### PR TITLE
Make SocketAddress Hashable.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

In general there's no good justification for `SocketAddress` not to be
`Hashable`: it's a well-formed value type that can easily have a
clearly-defined version of hashability defined. In particular, we have
a clear definition of equatability for this type already present in the
code.

This makes SocketAddress more broadly useful in a range of
circumstances.

Modifications:

- Defined `Hashable` for `SocketAddress`.

Result:

Users can obtain the hash value of `SocketAddress`.

### CODE REVIEW

File: Sources/NIO/SocketAddresses.swift [hunk 1]
```diff
@@ -363,6 +363,35 @@ extension SocketAddress: Equatable {
...
 
+/// We define an extension on `SocketAddress` that gives it an elementwise hashable conformance, using
+/// only the elements defined on the structure in their man pages (excluding lengths).
+extension SocketAddress: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        switch self {
+        case .unixDomainSocket(let uds):
+            hasher.combine(0)
+            hasher.combine(uds.address.sun_family)
+            withUnsafeBytes(of: uds.address.sun_path) {
```
Comment (weissi) [inline Sources/NIO/SocketAddresses.swift:374]: This is actually incorrect in both hashable and `==`. This needs to only compare the bytes until the first `0` byte I think.

Comment (weissi) [reply Sources/NIO/SocketAddresses.swift:374]: (ignore this is we know for sure that all trailing bytes are always `0` for example  in which case we should add an `assert`).

Comment (Lukasa) [reply Sources/NIO/SocketAddresses.swift:374]: Yeah, I felt pretty weird about it as I was writing it tbh.

Comment (Lukasa) [reply Sources/NIO/SocketAddresses.swift]: Updated, please take a look.

File: Sources/NIO/SocketAddresses.swift [hunk 2]
```diff
@@ -354,15 +354,62 @@ extension SocketAddress: Equatable {
...
-            return memcmp(&sunpath1, &sunpath2, MemoryLayout.size(ofValue: sunpath1)) == 0
+            let bufferSize = MemoryLayout.size(ofValue: addr1.address.sun_path)
+
+            // These uses of withMemoryRebound(to:) are fine: the `strncmp` call cannot possibly re-enter Swift code, and so we cannot possibly violate
+            // Swift's strict aliasing rules by way of this type-pun.
+            return withUnsafePointer(to: addr1.address.sun_path) { sunpath1 in
+                return withUnsafePointer(to: addr2.address.sun_path) { sunpath2 in
+                    return sunpath1.withMemoryRebound(to: Int8.self, capacity: bufferSize) { sunpath1 in
+                        return sunpath2.withMemoryRebound(to: Int8.self, capacity: bufferSize) { sunpath2 in
+                            return strncmp(sunpath1, sunpath2, MemoryLayout.size(ofValue: addr1.address.sun_path)) == 0
```
Comment (glbrntt) [inline Sources/NIO/SocketAddresses.swift:365]: Isn't this just another `bufferSize`?

**Comment by Lukasa** [comment]
Hit now-closed issue #554. @weissi are we sure that fix worked?

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

File: Sources/NIO/SocketAddresses.swift (hunk 2 above)
Comment (Lukasa) [reply Sources/NIO/SocketAddresses.swift:365]: Yup, I just missed this one.

**Comment by weissi** [comment]
@Lukasa pretty sure the original fix fixed one issue (which we hit frequently) but there might be a new one. Mind filing a fresh one with the printout?

**Comment by Lukasa** [comment]
See #1535.

**Pull request merged**